[#107] capacitySetting TipView 디자인 변경

Seon-U opened the pull request:
## 📝 작업 내용
> - TipView의 프로필 사진이 삭제되며, 데이터를 넘겨주던 뷰에서 이미지 변수들을 삭제함
> - TipView가 비디오 포맷따라 동적으로 변경되도록 Init을 통해 대부분의 변수를 초기화하도록 설정함
> - CapacitySettingView에서 팁뷰 버튼의 굵기를 semibold로 변경함
> - TipView를 전반적으로 다시 그림

### 스크린샷 (선택)
| 15 pro 기준 기본화질 화면 | 15 pro 기준 고화질 화면 | 아이패드 미니 6세대 기준 화면
|--|--|--|
| ![Simulator Screenshot - iPhone 15 Pro - 2024-09-15 at 21 03 38](https://github.com/user-attachments/assets/78bfa40b-495b-44c4-9c59-c9d45e027abd) | ![Simulator Screenshot - iPhone 15 Pro - 2024-09-15 at 21 03 23](https://github.com/user-attachments/assets/6550f395-2f75-44df-ad0a-41bdcfbfad35) | ![Simulator Screenshot - iPad mini (6th generation) - 2024-09-15 at 20 59 25](https://github.com/user-attachments/assets/e2df6013-71fd-4c01-916c-b5b0aaaf865f) |


## 💬 리뷰 요구사항(선택)

> videoFormat 기준으로 바뀌는 값들을 미리 선언하고, init을 통해 초기화함
> 일부는 padding값을, 일부는 Spacer를 사용하여 최대한 다른 기기에서도 잘리지 않도록 함

Seon-U committed:
feat: [#107] Change capacitySetting TipView

- TipView의 프로필 사진이 삭제되며, 데이터를 넘겨주던 뷰에서 이미지 변수들을 삭제함
- TipView가 비디오 포맷따라 동적으로 변경되도록 Init을 통해 대부분의 변수를 초기화하도록 설정함
- CapacitySettingView에서 팁뷰 버튼의 굵기를 semibold로 변경함
- TipView를 전반적으로 다시 그림

zaehorang commented on MoonCrystal/MoonCrystal/Sources/Views/MainHomeView/MainHomeView.swift:
```diff
@@ -107,11 +107,11 @@ struct MainHomeView: View {
                 } else if pathValue == "SettingView" {
                     
-                    CapacitySettingView(path: $navPath, totalCapacity: Int(totalCapacity.byteToGB()), videoFormat: selectedType!, favoriteIdol: userProfile.first?.favoriteIdol ?? "최애", profileImage: userProfile.first?.image)
+                    CapacitySettingView(path: $navPath, totalCapacity: Int(totalCapacity.byteToGB()), videoFormat: selectedType!, favoriteIdol: userProfile.first?.favoriteIdol ?? "최애")
```
현재 videoFormat을 강재 언래핑해서 사용하고 있는데 이 이유가 해당 변수에 디폴트 값이 들어 있어서 그런 거라면 해당 변수를 옵셔널 타입이 아닌 변수로 설정해보는 것도 고려해보면 좋을 거 같아요!

`@State var selectedType: VideoFormatCapacity? = .defaultQuality`
- 꼭 옵셔널 타입으로 선언해야 하는 이유가 있을까요? 나우가 짜신 코드가 아니라면 이후에 꾸기와 같이 얘기해봐요~

Seon-U replied:
제가 짠 코드가 아니라 확실한 건 아니지만, default값 때문에 강제 언래핑되는 게 아니라, 옵셔널 타입 때문인듯합니다.
- FormatInputView에서 라디오 버튼을 통해 선택하게 될 때 초기 상태에는 어떤 라디오 버튼도 선택되어 있지 않도록 설정을 하기 위해서 일부러 옵셔널로 선언되었다고 생각합니다. 
- 다만 다른 뷰에서는 format이 이미 고정된 플로우였기 때문에 이렇게 설정한 게 아닌가 생각합니다.

확실한 건 아니라, 자세한 내용은 꾸기와 이야기해야 할듯해요~

zaehorang replied:
고화질 상태에서 선택되지 않은 상태가 있어 옵셔널 타입으로 선언된 것을 확인했습니다!

위에서 제가 언급했던 강제 언래핑은 selectedType!에 대한 것이었습니다. 초기 값을 확실히 넣어주셨기 때문에, 이 부분은 간단한 코드 작성을 위해 아래와 같이 작성하셔도 문제 없을 것 같습니다 🐯
```
CapacitySettingView(path: $navPath, totalCapacity: Int(totalCapacity.byteToGB()), videoFormat: selectedType!, favoriteIdol: userProfile.first?.favoriteIdol ?? "최애")
```

zaehorang replied:
고화질 상태에서 선택되지 않은 상태가 있어 옵셔널 타입으로 선언된 것을 확인했습니다!

위에서 제가 언급했던 강제 언래핑은 selectedType!에 대한 것이었습니다. 초기 값을 확실히 넣어주셨기 때문에, 이 부분은 간단한 코드 작성을 위해 아래와 같이 작성하셔도 문제 없을 것 같습니다 🐯
```
CapacitySettingView(path: $navPath, totalCapacity: Int(totalCapacity.byteToGB()), videoFormat: selectedType!, favoriteIdol: userProfile.first?.favoriteIdol ?? "최애")
```

leesungkug replied:
와우 맞습니다 고화질 상태에서 선택되지 않은 상태가 nil값으로 설정되어있습니다. 늦게 답변을 드려서 죄송합니다 ㅜㅜ

zaehorang commented on MoonCrystal/MoonCrystal/Sources/Views/SelectCapacityView/TipView.swift:
```diff
@@ -7,21 +7,44 @@
-    let title = "정리를 시작하기 전,\n얼마나 삭제할 지 감이 안 온다면?"
-    var profileImage: Data?
+    
+    let title = "GB를 확보하려면\n얼마나 삭제해야 할까요?"
+    var videoFormat: VideoFormatCapacity
+    var videoFormatString: String
+    var standardGB: Int
+    var photoNumber: Int
+    var videoTime: Int
+    
+    /// 화면 초기화에서 모든 변수값 한번만 계산하도록 함
+    init(videoFormat: VideoFormatCapacity) {
+        self.videoFormat = videoFormat
+        
+        /// 기본화질과 고화질 구분
+        let isDefaultQuality = videoFormat == .defaultQuality
+        /// Byte 단위로 된 용량
+        let capacity = isDefaultQuality ? 5368709120 : 21474836480
+        
+        videoFormatString = isDefaultQuality ? "기본화질" : "고화질"
+        standardGB = isDefaultQuality ? 5 : 20
+        
+        photoNumber = MediaCapacityConverter.capacityToPhoto(capacity: capacity)
+        videoTime = MediaCapacityConverter.capacityToMinute(capacity: capacity, format: videoFormat)
+    }
```
현재 이 뷰는 **videoFormat**에 따라 표시되는 값들이 달라지는 것으로 보입니다.

1. 첫 번째로, 이 뷰에서 데이터를 처리하는 로직을 포함하는 것이 적절한지 고민이 됩니다. 저희 아키텍처가 **MV**이기 때문에, 데이터 처리 로직은 모델에 위치시키는 것이 더 적합하지 않을까 하는 생각이 듭니다. 이 부분에 대한 의견이 궁금합니다.
2. 두 번째로, 만약 로직이 간단해서 해당 뷰에서 처리하는 것이 맞다고 하더라도, 뷰에 표시될 변수들이 **var로 선언된 이유**가 궁금합니다. 값들이 변경될 필요가 없다고 생각되는데, let으로 선언할 수 있지 않을까 싶습니다. 이에 대한 설명을 부탁드립니다!

Seon-U replied:
일단,  전체 앱의 아키텍처를 고려하고 내린 판단은 아니었습니다. 다만 TipView자체가 자주 들어왔다 나가는 화면은 아니라고 판단했고, 따라서 외부에서 변화를 감지하여 추적하거나 (@StateObject로 관리) 별도 enum을 통해서 계산된 값을 화면에서 계속 호출하기보다는 화면에 들어올 때 한번에 처리하는 게 효율적이라고 생각했습니다.

그리고 var와 let의 경우 제가 개념이 잘못잡혀있어서 뷰 밖에서 값을 받을 때 var로 받는 게 습관이 되어있던 것 같습니다.
init 외에 화면에서 변동되는 점은 없기 때문에 let이 맞다고 생각합니다.
수정해서 PR 다시 올렸습니다~

zaehorang replied:
제가 전달한 MV 분리의 목적은 효율성보다는 코드 관리 측면에서 말씀드린 것이었습니다! 해당 로직을 모델에서 처리하든, 뷰에서 처리하든 성능상의 큰 차이는 없을 것으로 보입니다.

다만, 이후 파일이 커질 수 있다는 점을 감안했을 때, 저희 아키텍처의 방향성은 모델(특정 처리 로직, 비즈니스 로직 등을 다루는 클래스, 구조체 등)과 뷰를 명확히 분리하여, 뷰가 데이터의 처리 로직을 알지 않도록 하는 것이 중요합니다. 이는 저희 팀이 정한 약속이기도 합니다.

따라서 짧은 로직이라도 모델로 분리하는 것이, 이후 유지보수나 다른 팀원이 코드를 이해하기에 더 편리하고 일관된 방법이라고 생각했습니다!

Seon-U replied:
코드 관리 측면에서는 그게 맞는 듯합니다. 
- models에 TipViewModel을 추가하여 별도 class로 분리하였습니다.
- view init에서는 viewModel만 초기화하도록 설정했습니다.

zaehorang reviewed (commented):
코멘트 남겼습니다 🐯
추석 연휴 기간에도 고생이 많으십니다..!

Seon-U committed:
refactor: [#108] Change var to let in TipVIew

- TipView안에서 활용되는 var 변수들을 let 상수값으로 변경함

Seon-U committed:
refactor: [#107] Seperate View to ViewModel

- TipViewModel을 별도로 생성했습니다.
- init에서 처리하던 내용을 전부 TipViewModel에서 별도로 관리하게 하고, 그 값을 viewModel로 불러오도록 설정함

Seon-U committed:
refactor: [#107] Change class to struct

- class를 struct로 변경함
- 한 view에서만 지속적으로 활용되고, 상태관리가 필요없기에 struct가 더 적합하다고 판단함

leesungkug reviewed (approved):
추석연휴에 신경써주셔서 감사합니다!~

zaehorang commented on MoonCrystal/MoonCrystal/Sources/Models/TipViewModel.swift:
**View를 위한 모델을 따로 만드는 것보다는**, 뷰와 모델을 명확히 분리하여 **뷰는 모델이 어떤 일을 하는지 모르고**, 모델은 **뷰가 어떤 일을 하는지 모르게 만드는 것이 분리의 목적이라고 생각하는데 어떻게 생각하시나요?**.

해당 모델은 뷰에 **어떤 값을 보여줄지를 알려주는 역할**이기는 하지만, 이 모델을 **해당 뷰에서만 사용할 수 있게** 만들기보다는, **다른 곳에서도 재사용 가능**하도록 **더 일반적인 역할**을 하도록 분리하는 것이 좋을 것 같습니다. 예를 들어, **비디오 타입을 알려주면 얼만큼을 지워야 하는지를 계산해주는 모델**로 말이죠.

또한, 뷰에서 사용되는 값들(예: `title`, `videoFormattingString`)은 **뷰와 직접적으로 관련된 값**이기 때문에, **뷰에서 처리**하여 표시하는 것이 더 적절하다고 생각합니다.

이후 팁 뷰가 바뀌어도 해당 모델을 수정하지 않고 쓸 수있게 말이죠!

Seon-U replied:
제가 고민하고 있는 부분은 
- 첫째, 일단 팁뷰 외에서 이 모델이 사용될 가능성은 낮다고 생각합니다.
- 둘째, 비디오 타입을 알려주면 얼만큼 지워야 하는지 계산해주는 모델이라기엔 그 계산을 이미 MediaCapacityConverter class에서 해주고 있고 지금 이 **별도 모델에서 하는 일은 단순히 그 기준을 기본화질과 고화질이냐에 따라 5GB냐, 20GB냐 나눠주는 데 불과**하기 때문입니다. 그렇기에 저는 **애초에 이 일을 init으로 처리했고, 추후 모델 분리도 단순히 코드 관리를 효율적으로 하기 위해 나눠준 데**에 불과했습니다. 

뷰의 로직과 보여주는 것의 분리를 목적으로 화면을 나누기에는 이미 그 로직은 MediaCapacityConverter에서 처리하고 있고, 단순히 if, else로 "고화질이냐, 기본화질이냐" 기준을 추가하는 것에 불과하기 때문에 계산해주는 모델로 분리하는 것이 의미가 있다고 생각하기 어려운 듯합니다. 물론 이건 저 개인 생각이기에, 그럼에도 분리하는 것이 옳다고 생각하신다면 호랑의 의견에 따르겠습니다.

zaehorang replied:
우선 재사용성 측면을 고려하면, 다른 뷰에서 사용할 가능성을 말씀드렸는데, 현재 5GB, 20GB라는 값들이 변경될 가능성도 고려해볼 필요가 있을 것 같습니다.

말씀하신 대로, 이미 MediaCapacityConverter에서 대부분의 로직이 처리되고 있다면, 이 모델에서 별도로 처리하기보다는, 해당 모델 객체에 필요한 로직을 추가하는 것이 더 적합하다고 생각합니다. 예를 들어, 변환하고자 하는 GB와 format 타입을 파라미터로 받고, 뷰에서 필요한 값을 반환하는 메서드를 추가하는 방식이 될 수 있겠네요.

저도 전체적인 코드 구성을 완전히 보지 않고 말씀드리다 보니, 의견이 약간 바뀐 부분이 있습니다. 🥲

MV 아키텍처의 목적은 뷰와 모델을 명확히 분리하는 것인데, 현재 코드에서는 그 부분이 다소 모호하게 보이는 것 같습니다. 그래서 모델과 뷰의 책임을 더 명확히 분리하는 방법이 좋을 것 같다는 생각이 듭니다.

나우가 말씀하신 방향과 코드가 맞다고 생각되시면, 이전에 관련된 코드를 모두 나우가 작성하셨기 때문에 원하는 방향으로 진행하신 후 나중에 더 논의해봐도 좋을 것 같습니다!

Seon-U replied:
다시 한 번 commit 추가로 올렸습니다.
 - if-else 부분으로 갈리는 부분만 목표값 설정 struct로 별도로 설정하고, 다른 class를 활용하거나 단순 설정의 경우는 view에서 처리하도록 모델과 분리했습니다.
 - 다만 "변환하고자 하는 GB와 format 타입을 파라미터로 받고, 뷰에서 필요한 값을 반환하는 메서드를 추가"하는 방식은 이미 기존에 MediaConverter의 func에서 byte값을 받아 반환하고 있어 별도로 생성하진 않았습니다.
 
코멘트를 보면서 제 코드의 가독성이 나빴다는 걸 실감합니다. 바로 알아보기 쉽도록 다른 클래스를 참조하는 부분들은 조금 더 명확히 구분하고자 노력했고, if-else문 하나면 되는 간략한 부분도 조금 더 명확하게 struct로 설정했습니다. 여전히 MV 아키텍처에 대해서 잘 몰라 말하기는 조심스럽지만, 제가 판단하기에 수정이 필요하다고 생각하는 부분들은 어느정도 반영했습니다. 괜찮다면 이렇게 진행해도 될까요?

zaehorang reviewed (commented):
추석이라 바쁘실텐데,,,
계속된 피드백 반영 감사드립니다!

Seon-U committed:
refactor: [#107] Change viewModel to deleteGoal only

- 뷰모델을 deleteGoalTip으로 바꾸면서 삭제 용량 기준점만 반환하도록 설정함
- photoNumber 등 별도 class 함수로 계산되는 부분을 분리함
- title 등 변하지 않는 값은 뷰 안에서 기본값으로 설정함

zaehorang commented on MoonCrystal/MoonCrystal/Sources/Models/DeleteGoalTip.swift:
```diff
@@ -0,0 +1,22 @@
+    init(videoFormat: VideoFormatCapacity) {
+        /// 기본화질과 고화질 구분
+        let isDefaultQuality = videoFormat == .defaultQuality
+        videoFormatString = isDefaultQuality ? "기본화질" : "고화질"
```
이 부분만 뷰에서 처리하면 좋을 거 같아요!
고생하셨습니다🐯

Seon-U replied:
뷰와 관련된 텍스트는 맞지만 용량 기준점인 기본화질, 고화질이냐에 따라 나뉘는 텍스트이기도 합니다. 
기준을 정한다는 점에서 goal과 함께 두었습니다.
뷰에서 처리하는 게 크게 어려운 일은 아니지만, 나중에 용량기준이 조금 더 세분화된다면 이 부분 역시 수정된다고 생각합니다.
어떻게 생각하시나요?

zaehorang replied:
넵, 우선 이렇게 진행하시죠! 🐯
이후에 해당 데이터 타입 관련 모델을 리팩토링하면서 더 개선할 부분이 있으면 그때 논의해봐도 좋을 것 같습니다~

zaehorang reviewed (approved):
하나의 수정만 더 남겼습니다~
(뷰와 관련된 텍스트인 거 같아서요..!)
의견 있으면 남겨주시고 승인 누릅니다.👍
